kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+  end
+
+  -- Guard against queue shrinkage during handler invocation by using match.min below.
+  for _ = 1, math.min(entry_count, self:count()) do
+    self:delete_frontmost_entry()
+  end
+  if self.queue_full then
+    self:log(INFO, 'queue resumed processing')
+    self.queue_full = false
+  end
+end
+
+
+-- This function retrieves the queue parameters from a plugin configuration, converting legacy parameters
+-- to their new locations
+function Queue.get_params(config)
+  local queue_config = config.queue or {}
+  -- Common queue related legacy parameters
+  if config.retry_count and config.retry_count ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `retry_count` parameter in plugin %s ignored",
+      kong.plugin.get_id()))
+  end
+  if config.queue_size and config.queue_size ~= 1 and config.queue_size ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `queue_size` parameter in plugin %s converted to `queue.batch_max_size`",
+      kong.plugin.get_id()))
+    queue_config.batch_max_size = config.queue_size
+  end
+  if config.flush_timeout and config.flush_timeout ~= 2 and config.flush_timeout ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `flush_timeout` parameter in plugin %s converted to `queue.max_delay`",
+      kong.plugin.get_id()))
+    queue_config.max_delay = config.flush_timeout
+  end
+  -- Queue related opentelemetry plugin parameters
+  if config.batch_span_count and config.batch_span_count ~= 200 and config.batch_span_count ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `batch_span_count` parameter in plugin %s converted to `queue.batch_max_size`",
+      kong.plugin.get_id()))
+    queue_config.batch_max_size = config.batch_span_count
+  end
+  if config.batch_flush_delay and config.batch_flush_delay ~= 3 and config.batch_flush_delay ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `batch_flush_delay` parameter in plugin %s converted to `queue.max_delay`",
+      kong.plugin.get_id()))
+    queue_config.max_delay = config.batch_flush_delay
+  end
+
+  if not queue_config.name then
+    queue_config.name = kong.plugin.get_id()
+  end
+  return queue_config
+end
+
+
+-------------------------------------------------------------------------------
+-- Drain the queue, used during orderly shutdown and for testing.
+-- @param self Queue
+function Queue:_drain()
+end
+
+
+-------------------------------------------------------------------------------
+-- Add entry to the queue
+-- @param conf plugin configuration of the plugin instance that caused the item to be queued
+-- @param entry the value included in the queue. It can be any Lua value besides nil.
+-- @return true, or nil and an error message.
+function Queue:_enqueue(entry)
+  if entry == nil then
+    return nil, "entry must be a non-nil Lua value"
+  end
+
+  if self:count() >= self.capacity * Queue.CAPACITY_WARNING_THRESHOLD then
```

Comment:
RHS could probably be computed and stored.
I'm not completely about the impact it would have on the proxy path but given it is called for each enqueue operation, it seems like a worthwhile optimization.
What do you think?

Reply:
I don't think it will be noticeable at all because the arithmetic operation is likely drowned in cache misses and memory accesses.  I'm very reluctant to add such micro-optimizations without knowing that they have a measureable impact as they usually make the code more complex.